components/support/sync15-traits/src/telemetry.rs

```diff
@@ -0,0 +1,777 @@
+    #[inline]
+    pub fn new_failed(&mut self, n: u32) {
+        self.new_failed += n;
+    }
+
+    /// Increment the value of `reconciled` by `n`.
+    #[inline]
+    pub fn reconciled(&mut self, n: u32) {
+        self.reconciled += n;
+    }
+
+    /// Get the value of `applied`. Mostly useful for testing.
+    #[inline]
+    pub fn get_applied(&self) -> u32 {
+        self.applied
+    }
+
+    /// Get the value of `failed`. Mostly useful for testing.
+    #[inline]
+    pub fn get_failed(&self) -> u32 {
+        self.failed
+    }
+
+    /// Get the value of `new_failed`. Mostly useful for testing.
+    #[inline]
+    pub fn get_new_failed(&self) -> u32 {
+        self.new_failed
+    }
+
+    /// Get the value of `reconciled`. Mostly useful for testing.
+    #[inline]
+    pub fn get_reconciled(&self) -> u32 {
+        self.reconciled
+    }
+}
+
+/// Outgoing record for an engine's sync
+#[derive(Debug, Default, Serialize)]
+pub struct EngineOutgoing {
+    #[serde(skip_serializing_if = "crate::skip_if_default")]
+    sent: usize,
+
+    #[serde(skip_serializing_if = "crate::skip_if_default")]
+    failed: usize,
+}
+
+impl EngineOutgoing {
+    pub fn new() -> Self {
+        EngineOutgoing {
+            ..Default::default()
+        }
+    }
+
+    #[inline]
+    pub fn sent(&mut self, n: usize) {
+        self.sent += n;
+    }
+
+    #[inline]
+    pub fn failed(&mut self, n: usize) {
+        self.failed += n;
+    }
+}
+
+/// One engine's sync.
+#[derive(Debug, Serialize)]
+pub struct Engine {
+    name: String,
+
+    #[serde(flatten)]
+    when_took: Stopwatch,
+
+    #[serde(skip_serializing_if = "EngineIncoming::is_empty")]
+    incoming: Option<EngineIncoming>,
+
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    outgoing: Vec<EngineOutgoing>, // one for each batch posted.
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(rename = "failureReason")]
+    failure: Option<SyncFailure>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    validation: Option<Validation>,
+}
+
+impl Engine {
+    pub fn new(name: impl Into<String>) -> Self {
+        Self {
+            name: name.into(),
+            when_took: Stopwatch::new(),
+            incoming: None,
+            outgoing: Vec::new(),
+            failure: None,
+            validation: None,
+        }
+    }
+
+    pub fn incoming(&mut self, inc: EngineIncoming) {
+        assert!(self.incoming.is_none());
+        self.incoming = Some(inc);
+    }
+
+    pub fn outgoing(&mut self, out: EngineOutgoing) {
+        self.outgoing.push(out);
+    }
+
+    pub fn failure(&mut self, err: impl Into<SyncFailure>) {
+        // Currently we take the first error, under the assumption that the
+        // first is the most important and all others stem from that.
+        let failure = err.into();
+        if self.failure.is_none() {
+            self.failure = Some(failure);
+        } else {
+            log::warn!(
```

Comment:
Hmmm, side note—we're going to need to figure out how to get these logs out on Desktop. We do vendor the `log` crate, but I think it only logs to stdout—and, since the logger is static, we'd need to figure out how to send some of this stuff to the Sync log, but not everything (like, say, WebRender or Stylo logs). Dogear gets around this by reimplementing the logging macros itself, and making you pass the driver as the first argument, but that doesn't really work here.

More food for thought than anything at this point.

Reply:
Yeah... Not sure. I don't think it's plausible for us to thread logging everywhere we'd need it the way dogear does. We could always do something like mimic the `log` crate, just to allow it to be separately configured though.